### PR TITLE
Indicate source of element for completion

### LINKED ISSUE #210: Mark element coming from XML Schema/DTD for completion
While studying bug with duplicate elements https://github.com/redhat-developer/vscode-xml/issues/78#issuecomment-437049479 I find it should be good to mark the completion elements origin (from XML Schema, from DTD, copy sibling)

A quick idea is to add -  XML Schema:

image

@fbricon @NikolasKomonen @ralfhandl what do you think about that?

### PR DESCRIPTION
**Pull request by NikolasKomonen**
Fixes #210

Example with A description and filename

Selection_045

### CODE REVIEW

**Comment by NikolasKomonen** [comment]
@fbricon @angelozerr 

Thoughts on the final result, any preferences for me to change anything?

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java [hunk 1]
```diff
@@ -84,25 +94,51 @@ public void onTagOpen(ICompletionRequest request, ICompletionResponse response)
...
-			String p, boolean forceUseOfPrefix, ICompletionRequest request, ICompletionResponse response)
+			String p, boolean forceUseOfPrefix, ICompletionRequest request, ICompletionResponse response, String schemaURI)
 			throws BadLocationException {
 		XMLGenerator generator = request.getXMLGenerator();
 		for (CMElementDeclaration child : cmElements) {
 			String prefix = forceUseOfPrefix ? p : (element != null ? element.getPrefix(child.getNamespace()) : null);
-			String label = child.getName(prefix);
+			String fileTypeLabel = "";
+			// if(schemaURI != null) {
+			// 	if(schemaURI.endsWith("xsd")) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java]: no commented code please

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java [hunk 2]
```diff
@@ -84,25 +94,51 @@ public void onTagOpen(ICompletionRequest request, ICompletionResponse response)
...
+			// 		fileTypeLabel = " - XSD";
+			// 	}
+			// 	else if (schemaURI.endsWith("dtd")) {
+			// 		fileTypeLabel = " - DTD";
+			// 	}
+			// }
+			String label = child.getName(prefix) + fileTypeLabel;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java]: useless concatenation

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java [hunk 3]
```diff
@@ -84,25 +94,51 @@ public void onTagOpen(ICompletionRequest request, ICompletionResponse response)
...
 			CompletionItem item = new CompletionItem(label);
 			item.setFilterText(request.getFilterForStartTagName(label));
 			item.setKind(CompletionItemKind.Property);
-			String documentation = child.getDocumentation();
-			if (documentation != null) {
+			StringBuilder sb = new StringBuilder();
+			String documentation;
+			String tempDoc = child.getDocumentation();
+			boolean tempDocHasContent = tempDoc != null && !tempDoc.isEmpty();
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java]: there's a StringUtils.isEmpty method

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java [hunk 4]
```diff
@@ -84,25 +94,51 @@ public void onTagOpen(ICompletionRequest request, ICompletionResponse response)
...
+			if(tempDocHasContent) {
+				sb.append(tempDoc);
+			}
+			
+			if(schemaURI != null) {
+				if(tempDocHasContent) {
+					String lineSeparator = System.getProperty("line.separator");
+					sb.append(lineSeparator + lineSeparator);
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java]: don't concatenate, append twice

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java [hunk 5]
```diff
@@ -84,25 +94,51 @@ public void onTagOpen(ICompletionRequest request, ICompletionResponse response)
...
+				}
+				Path schemaPath = Paths.get(schemaURI);
+				sb.append("File: ");
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java]: Source:

**Comment by fbricon** [comment]
Missing unit tests for xsd and dtd sources

**Comment by fbricon** [comment]
Overall it looks pretty good (using Source: instead of File:)
May-27-2019 15-57-39

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/model/CMDocument.java [hunk 6]
```diff
@@ -32,4 +32,8 @@
 	 */
 	CMElementDeclaration findCMElement(DOMElement element, String namespace);
 
+	String getURI();
+
+	String setURI(String uri);
```
Comment (angelozerr) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/model/CMDocument.java]: IMHO this setUri should not exists and uri should be initialized in the constructor of each CMDocument implementation.

Comment (NikolasKomonen) [reply org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/model/CMDocument.java]: Alright, will update that.

**Comment by NikolasKomonen** [comment]
@fbricon Updated with tests, now 'Source' is used instead of 'File' and the file URI is provided in the constructor as @angelozerr indicated.

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/CMXSDDocument.java [hunk 7]
```diff
@@ -43,11 +43,24 @@
...
 	private Collection<CMElementDeclaration> elements;
 
+	private String uri;
+
 	public CMXSDDocument(XSModel model) {
 		this.model = model;
 		this.elementMappings = new HashMap<>();
 	}
 
+	public CMXSDDocument(XSModel model, String uri) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/CMXSDDocument.java]:
```java
public CMXSDDocument(XSModel model, String uri) {
   this(model);
   this.uri = uri;
}
```

**Comment by fbricon** [comment]
Please fix the conflict with master

File: org.eclipse.lsp4xml/src/test/resources/xsd/invoice.xsd [hunk 8]
```diff
@@ -8,9 +8,13 @@
     		<xsd:documentation>An invoice type...</xsd:documentation>
     	</xsd:annotation>
     	<xsd:sequence>
-    		<xsd:element name="date" type="xsd:date" nillable="true"></xsd:element>
-            <xsd:element name="number" type="xsd:int"></xsd:element>
-            <xsd:element name="products" type="productsType"></xsd:element>
+    		<xsd:element name="date" type="xsd:date" nillable="true">
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/test/resources/xsd/invoice.xsd]: please fix formatting

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/CMXSDContentModelProvider.java [hunk 9]
```diff
@@ -73,7 +73,8 @@ public CMDocument createCMDocument(String key) {
 		XSModel model = getLoader().loadURI(key);
 		if (model != null) {
 			// XML Schema can be loaded
-			return new CMXSDDocument(model);
+			CMXSDDocument document = new CMXSDDocument(model, key);
+			return document;
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/xsd/contentmodel/CMXSDContentModelProvider.java]: return new CMXSDDocument(model, key);

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/dtd/contentmodel/CMDTDDocument.java [hunk 10]
```diff
@@ -58,6 +68,15 @@
 		return elements;
 	}
 
+	@Override
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/dtd/contentmodel/CMDTDDocument.java]: move annotation after javadoc

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/dtd/contentmodel/CMDTDDocument.java [hunk 11]
```diff
@@ -41,6 +41,16 @@
 	private List<CMElementDeclaration> elements;
 	private DTDGrammar grammar;
 	private List<String> hierachies;
+	private String uri;
+
+
+	public CMDTDDocument() {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/dtd/contentmodel/CMDTDDocument.java]: remove empty lines

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java [hunk 12]
```diff
@@ -101,8 +112,26 @@ private void fillWithChildrenElementDeclaration(DOMElement element, Collection<C
...
+					String lineSeparator = System.getProperty("line.separator");
+					sb.append(lineSeparator);
+					sb.append(lineSeparator);
+				}
+				Path schemaPath = Paths.get(schemaURI);
+				sb.append("Source: ");
+				sb.append(schemaPath.getFileName().toString());
+			}
+			documentation = sb.toString();
+			if (documentation != null && !documentation.isEmpty()) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java]: documentation can't be null at this point

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java [hunk 13]
```diff
@@ -101,8 +112,26 @@ private void fillWithChildrenElementDeclaration(DOMElement element, Collection<C
...
+				if(tempDocHasContent) {
+					String lineSeparator = System.getProperty("line.separator");
+					sb.append(lineSeparator);
+					sb.append(lineSeparator);
+				}
+				Path schemaPath = Paths.get(schemaURI);
+				sb.append("Source: ");
+				sb.append(schemaPath.getFileName().toString());
+			}
+			documentation = sb.toString();
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/extensions/contentmodel/participants/ContentModelCompletionParticipant.java]: String documentation = sb.toString();

**Comment by NikolasKomonen** [comment]
@fbricon Updated

**Pull request merged**